src/hotspot/cpu/x86/interp_masm_x86.cpp

```diff
@@ -1184,6 +1184,13 @@ void InterpreterMacroAssembler::remove_activation(
 #else
     // Load fields from a buffered value with an inline class specific handler
     load_klass(rdi, rax, rscratch1);
+    movptr(rscratch1, rax);
+    // Skip scalarization for vector value objects (concrete vectors and payloads).
+    super_call_VM_leaf(CAST_FROM_FN_PTR(address, SharedRuntime::is_vector_value_instance), rdi);
```

Comment:
> I'm curious what's the current way to disable scalarization for return values of primitive types other than vector. Or do all primitive classes unconditionally scalarized irrespective of how many elements they comprise?
> 
Currently entire handling operates under two global flags `InlineTypeReturnedAsFields` and `InlineTypePassFieldsAsArgs`
I introduced special handling to prevent scalarization for vectors in `InlineKlass::can_be_passed_as_fields` and `InlineKlass::can_be_returned_as_fields`.

> It makes sense to introduce a flag on a Klass which signals when return value scalarization should be turned off (unless there's already such a flag present).

Agree, I can handle this on lworld branch since it will be good to enhance core infrastructure to disable scalarization per class basis even though value type semantics always enforces scalarization, vector values are special since multifield needs to be passed using vector and we do not support a mixed GPR+VECTOR calling convention currently.
 
> 
> I see `InlineKlass::can_be_returned_as_fields()` but don't fully understand how it interacts with interpreter.

InlineKlass contains unpack / pack handlers , interpreter and c1 side handling directly operate under influence of above mentioned global flags. There is no mechanism to prevent scalarization for value objects currently, apart from c2 all the other execution engines uses pack/unpack handlers to pass / receive a value object.